### ISSUE TITLE
`rule-name`: `no-array-callback-reference` option to allow functions defined in module

### ISSUE THREAD

**Issue by noahbald** [sindresorhus/eslint-plugin-unicorn#2632]
### Description

I find it common to have a function written specifically for use as a callback and I find it frustrating to have to create a redundant wrapper for it, especially when using it as a type guard. I get the motivation for this rule, but there shouldn't be much risk of the function interface unexpectedly changing when it's written specifically for it's use.

Is it possible to add an option to the config so that we can use callback references when it's defined in the same module as it's usage?

### Examples

```js
// ❌ Error if callback is from module
import { process } from "./module";

[0, 1, 2].forEach(process);

// ✅ Allow if process is defined locally
const process = (item) => console.log(item);

[0, 1, 2].forEach(process);
```

```ts
// ❌ Error if callback is from module
import { isNumber } from "./module";

const odds = [1, 3, "5"];
const evens = [0, 2, 4];
if (odds.every(isNumber) && evens.every(isNumber)) {
  const myNumber = evens[0] || odds[0];
}

// ✅ Allow even when used multiple times
const isNumber = (item: unknown): item is number => typeof item === "number";

const odds = [1, 3, "5"];
const evens = [0, 2, 4];
if (odds.every(isNumber) && evens.every(isNumber)) {
  const myNumber = evens[0] || odds[0];
}
```

**Comment by osdiab** [comment]
Or if you could specifically whitelist certain callbacks, e.g. like the above we have useful type guards for working with arrays.

**Comment by sindresorhus** [comment]
I’d prefer not to add this option. The rule intentionally reports locally declared callbacks too, because the issue is the implicit iterator arguments, not only external API churn. Type predicate callbacks are already allowed for the array methods where wrapping them would lose TypeScript narrowing. For intentional one-off cases, an inline disable is clearer than adding another configuration path.
